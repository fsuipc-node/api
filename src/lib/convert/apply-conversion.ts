import { VM } from 'vm2';

import { Offset } from '@shared/offset';
import { RawOffsetValue } from '@shared/offset-values';
import { ConvertedOffsetValue } from '@shared/converted-offset-values';

export const applyConversion = (offset: Offset, rawOffsetValue: RawOffsetValue): ConvertedOffsetValue => {
  if (!offset.convert) {
    return rawOffsetValue;
  }

  if (offset.hasMapping) {
    if (offset.isMappingInvalid) {
      return 'INVALID_MAPPING_FUNCTION';
    }

    return offset.mappingFunction(rawOffsetValue);
  } else {
    // tslint:disable-next-line:no-eval
    if (offset.isInvalidConvertExpression) {
      return 'UNSUPPORTED_CONVERSION_EXPRESSION';
    }

    return new VM().run(
      replaceOffsetExpressionValue(offset, rawOffsetValue)
    );
  }
};

export const replaceOffsetExpressionValue = (offset: Offset, rawOffsetValue: RawOffsetValue): string => {
  return offset.convert.replace(
    new RegExp(/{VAL}/g),
    Array.isArray(rawOffsetValue)
      ? `${JSON.stringify(rawOffsetValue)}`
      : typeof rawOffsetValue === 'string'
        ? `'${rawOffsetValue}'`
        : rawOffsetValue.toString()
  );
};
