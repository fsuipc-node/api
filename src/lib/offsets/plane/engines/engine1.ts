import { Type } from 'fsuipc';

import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';
import { OffsetCategory } from '@shared/offset-category';

export const engine1: OffsetList = {
  engine1ThrottleLever: new Offset({
    value: 0x88C,
    name: 'engine1ThrottleLever',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 throttle lever - percent - negative = reverse',
    convert: 'Math.round({VAL} < 0 ? {VAL} / 4096 * 100 : {VAL} / 16384 * 100)',
    type: Type.Int16,
    permission: 'r',
  }),
  engine1PropLever: new Offset({
    value: 0x88E,
    name: 'engine1PropLever',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 prop lever - percent - negative = reverse',
    convert: 'Math.round({VAL} < 0 ? {VAL} / 4096 * 100 : {VAL} / 16384 * 100)',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine1MixtureLever: new Offset({
    value: 0x890,
    name: 'engine1MixtureLever',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 prop lever - percent',
    convert: '{VAL} / 16384 * 100',
    type: Type.UInt16,
    permission: 'rw',
  }),
  engine1StarterSwitchPosition: new Offset({
    value: 0x892,
    name: 'engine1StarterSwitchPosition',
    category: OffsetCategory.ENGINE,
    description: 'engine1 start switch position - JET: 0=off 1=start 2=gen/alt - PROP: 0=off 1=right 2=left 3=both 4=start',
    type: Type.UInt16,
    permission: 'rw',
  }),
  engine1Firing: new Offset({
    value: 0x894,
    name: 'engine1Firing',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 combustion',
    convert: '!!{VAL}',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine1N2: new Offset({
    value: 0x896,
    name: 'engine1N2',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 N2 - also helo RPM %',
    convert: '{VAL} / 16384 * 100',
    type: Type.UInt16,
    permission: 'rw',
  }),
  engine1N1: new Offset({
    value: 0x898,
    name: 'engine1N1',
    category: OffsetCategory.ENGINE,
    description: 'engine 1 N1 - also helo RPM %',
    convert: '{VAL} / 16384 * 100',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine1PropRPM: new Offset({
    value: 0x898,
    name: 'engine1PropRPM',
    category: OffsetCategory.ENGINE,
    description: 'derive RPM by multiplying by  - negative = counter-rotating propeller',
    convert: '{VAL} / 16384 / 65536',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine1ThrottleLeverControl: new Offset({
    value: 0x89A,
    name: 'engine1ThrottleLeverControl',
    category: OffsetCategory.ENGINE,
    description: 'engine1 throttle lever control - -4096 to +16384',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine1FuelFlowLbHourSSL: new Offset({
    value: 0x8A0,
    name: 'engine1FuelFlowLbHourSSL',
    category: OffsetCategory.ENGINE,
    description: 'engine1 fuel flow (lb per hour, sea level)',
    convert: '{VAL} / 128',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine1AntiIce: new Offset({
    value: 0x8B2,
    name: 'engine1AntiIce',
    category: OffsetCategory.ENGINE,
    description: 'engine1 anti-ice/carb heat active',
    convert: '!!{VAL}',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine1OilTemp: new Offset({
    value: 0x8B8,
    name: 'engine1OilTemp',
    category: OffsetCategory.ENGINE,
    description: 'engine1 oil temp - celsius',
    convert: 'Math.round({VAL} * 140 / 16384)',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine1OilPres: new Offset({
    value: 0x8BA,
    name: 'engine1OilPres',
    category: OffsetCategory.ENGINE,
    description: 'engine1 oil pressure - PSI',
    convert: 'Math.round({VAL} * 55 / 16384)',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine1PressureRatio: new Offset({
    value: 0x8BC,
    name: 'engine1PressureRatio',
    category: OffsetCategory.ENGINE,
    description: 'engine1 pressure ratio',
    convert: '{VAL} * 1.6 / 16384',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine1EGT: new Offset({
    value: 0x8BE,
    name: 'engine1EGT',
    category: OffsetCategory.ENGINE,
    description: 'engine1 exhaust gas temperature - Trust only on jet engine - celsius',
    convert: 'Math.round({VAL} * 860 / 16384)',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine1MP: new Offset({
    value: 0x8C0,
    name: 'engine1MP',
    category: OffsetCategory.ENGINE,
    description: 'engine1 manifold pressure - inHg',
    convert: '{VAL} / 1024',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine1RPMScaler: new Offset({
    value: 0x8C8,
    name: 'engine1RPMScaler',
    category: OffsetCategory.ENGINE,
    description: 'engine1 RPM Scaler - props: used to calculate RPM',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine1OilQuantity: new Offset({
    value: 0x8D0,
    name: 'engine1OilQuantity',
    category: OffsetCategory.ENGINE,
    description: 'engine1 oil quantity - percent',
    convert: 'Math.round({VAL} / 16384 * 100)',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine1Vibration: new Offset({
    value: 0x8D4,
    name: 'engine1Vibration',
    category: OffsetCategory.ENGINE,
    description: 'engine1 vibration',
    convert: 'Math.round({VAL} * 5 / 16384)',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine1HydPres: new Offset({
    value: 0x8D8,
    name: 'engine1HydPres',
    category: OffsetCategory.ENGINE,
    description: 'engine1 hydraulic pressure - PSI',
    convert: '{VAL} / 4',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine1HydQuantity: new Offset({
    value: 0x8DC,
    name: 'engine1HydQuantity',
    category: OffsetCategory.ENGINE,
    description: 'engine1 hydraulic quantity - PSI',
    convert: '{VAL} / 16384 * 100',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine1CHT: new Offset({
    value: 0x8E8,
    name: 'engine1CHT',
    category: OffsetCategory.ENGINE,
    description: 'engine1 cylinder head temperature - F',
    type: Type.Double,
    permission: 'r',
  }),
  engine1ITT: new Offset({
    value: 0x8F0,
    name: 'engine1ITT',
    category: OffsetCategory.ENGINE,
    description: 'engine1 turbine temperature - C',
    convert: '{VAL} / 16384',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine1Torque: new Offset({
    value: 0x8F4,
    name: 'engine1Torque',
    category: OffsetCategory.ENGINE,
    description: 'engine1 torque - trust only for helo - percent',
    convert: '{VAL} / 16384 * 100',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine1FuelPres: new Offset({
    value: 0x8F8,
    name: 'engine1FuelPres',
    category: OffsetCategory.ENGINE,
    description: 'engine1 fuel pressure - PSI',
    convert: '{VAL} / 144',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine1FuelUsedSinceStart: new Offset({
    value: 0x90C,
    name: 'engine1FuelUsedSinceStart',
    category: OffsetCategory.ENGINE,
    description: 'engine1 used fuel since start - lb',
    type: Type.Single,
    permission: 'r',
  }),
  engine1FuelElapsedTime: new Offset({
    value: 0x910,
    name: 'engine1FuelElapsedTime',
    category: OffsetCategory.ENGINE,
    description: 'engine1 fuel elapsed time - h',
    type: Type.Single,
    permission: 'r',
  }),
  engine1FuelFlowLbHour: new Offset({
    value: 0x918,
    name: 'engine1FuelFlowLbHour',
    category: OffsetCategory.ENGINE,
    description: 'engine1 fuel elapsed time - lb/h',
    type: Type.Double,
    permission: 'r',
  }),
  engine1TorqueFtLb: new Offset({
    value: 0x920,
    name: 'engine1TorqueFtLb',
    category: OffsetCategory.ENGINE,
    description: 'engine1 fuel elapsed time - ft lb',
    type: Type.Single,
    permission: 'r',
  }),
};
