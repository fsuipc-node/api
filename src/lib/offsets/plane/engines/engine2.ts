import { Type } from 'fsuipc';

import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';
import { OffsetCategory } from '@shared/offset-category';

export const engine2: OffsetList = {
  engine2ThrottleLever: new Offset({
    value: 0x924,
    name: 'engine2ThrottleLever',
    category: OffsetCategory.ENGINE,
    description: 'engine 2 throttle lever - percent - negative = reverse',
    convert: 'Math.round({VAL} < 0 ? {VAL} / 4096 * 100 : {VAL} / 16384 * 100)',
    type: Type.Int16,
    permission: 'r',
  }),
  engine2PropLever: new Offset({
    value: 0x926,
    name: 'engine2PropLever',
    category: OffsetCategory.ENGINE,
    description: 'engine 2 prop lever - percent - negative = reverse',
    convert: 'Math.round({VAL} < 0 ? {VAL} / 4096 * 100 : {VAL} / 16384 * 100)',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine2MixtureLever: new Offset({
    value: 0x928,
    name: 'engine2MixtureLever',
    category: OffsetCategory.ENGINE,
    description: 'engine 2 prop lever - percent',
    convert: '{VAL} / 16384 * 100',
    type: Type.UInt16,
    permission: 'rw',
  }),
  engine2StarterSwitchPosition: new Offset({
    value: 0x92A,
    name: 'engine2StarterSwitchPosition',
    category: OffsetCategory.ENGINE,
    description: 'engine2 start switch position - JET: 0=off 1=start 2=gen/alt - PROP: 0=off 1=right 2=left 3=both 4=start',
    type: Type.UInt16,
    permission: 'rw',
  }),
  engine2Firing: new Offset({
    value: 0x92C,
    name: 'engine2Firing',
    category: OffsetCategory.ENGINE,
    description: 'engine 2 combustion',
    convert: '!!{VAL}',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine2N2: new Offset({
    value: 0x92E,
    name: 'engine2N2',
    category: OffsetCategory.ENGINE,
    description: 'engine 2 N2 - also helo RPM %',
    convert: '{VAL} / 16384 * 100',
    type: Type.UInt16,
    permission: 'rw',
  }),
  engine2N1: new Offset({
    value: 0x930,
    name: 'engine2N1',
    category: OffsetCategory.ENGINE,
    description: 'engine 2 N1 - also helo RPM %',
    convert: '{VAL} / 16384 * 100',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine2PropRPM: new Offset({
    value: 0x930,
    name: 'engine2PropRPM',
    category: OffsetCategory.ENGINE,
    description: 'derive RPM by multiplying by  - negative = counter-rotating propeller',
    convert: '{VAL} / 16384 / 65536',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine2ThrottleLeverControl: new Offset({
    value: 0x932,
    name: 'engine2ThrottleLeverControl',
    category: OffsetCategory.ENGINE,
    description: 'engine2 throttle lever control - -4096 to +16384',
    type: Type.Int16,
    permission: 'rw',
  }),
  engine2FuelFlowLbHourSSL: new Offset({
    value: 0x938,
    name: 'engine2FuelFlowLbHourSSL',
    category: OffsetCategory.ENGINE,
    description: 'engine2 fuel flow (lb per hour, sea level)',
    convert: '{VAL} / 128',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine2AntiIce: new Offset({
    value: 0x94A,
    name: 'engine2AntiIce',
    category: OffsetCategory.ENGINE,
    description: 'engine2 anti-ice/carb heat active',
    convert: '!!{VAL}',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine2OilTemp: new Offset({
    value: 0x950,
    name: 'engine2OilTemp',
    category: OffsetCategory.ENGINE,
    description: 'engine2 oil temp - celsius',
    convert: 'Math.round({VAL} * 140 / 16384)',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine2OilPres: new Offset({
    value: 0x952,
    name: 'engine2OilPres',
    category: OffsetCategory.ENGINE,
    description: 'engine2 oil pressure - PSI',
    convert: 'Math.round({VAL} * 55 / 16384)',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine2PressureRatio: new Offset({
    value: 0x954,
    name: 'engine2PressureRatio',
    category: OffsetCategory.ENGINE,
    description: 'engine2 pressure ratio',
    convert: '{VAL} * 1.6 / 16384',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine2EGT: new Offset({
    value: 0x956,
    name: 'engine2EGT',
    category: OffsetCategory.ENGINE,
    description: 'engine2 exhaust gas temperature - Trust only on jet engine - celsius',
    convert: 'Math.round({VAL} * 860 / 16384)',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine2MP: new Offset({
    value: 0x958,
    name: 'engine2MP',
    category: OffsetCategory.ENGINE,
    description: 'engine2 manifold pressure - inHg',
    convert: '{VAL} / 1024',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine2RPMScaler: new Offset({
    value: 0x960,
    name: 'engine2RPMScaler',
    category: OffsetCategory.ENGINE,
    description: 'engine2 RPM Scaler - props: used to calculate RPM',
    type: Type.UInt16,
    permission: 'r',
  }),
  engine2OilQuantity: new Offset({
    value: 0x968,
    name: 'engine2OilQuantity',
    category: OffsetCategory.ENGINE,
    description: 'engine2 oil quantity - percent',
    convert: 'Math.round({VAL} / 16384 * 100)',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine2Vibration: new Offset({
    value: 0x96C,
    name: 'engine2Vibration',
    category: OffsetCategory.ENGINE,
    description: 'engine2 vibration',
    convert: 'Math.round({VAL} * 5 / 16384)',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine2HydPres: new Offset({
    value: 0x970,
    name: 'engine2HydPres',
    category: OffsetCategory.ENGINE,
    description: 'engine2 hydraulic pressure - PSI',
    convert: '{VAL} / 4',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine2HydQuantity: new Offset({
    value: 0x974,
    name: 'engine2HydQuantity',
    category: OffsetCategory.ENGINE,
    description: 'engine2 hydraulic quantity - percent',
    convert: '{VAL} / 16384 * 100',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine2CHT: new Offset({
    value: 0x980,
    name: 'engine2CHT',
    category: OffsetCategory.ENGINE,
    description: 'engine2 cylinder head temperature - F',
    type: Type.Double,
    permission: 'r',
  }),
  engine2ITT: new Offset({
    value: 0x988,
    name: 'engine2ITT',
    category: OffsetCategory.ENGINE,
    description: 'engine2 turbine temperature - C',
    convert: '{VAL} / 16384',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine2Torque: new Offset({
    value: 0x98C,
    name: 'engine2Torque',
    category: OffsetCategory.ENGINE,
    description: 'engine2 torque - trust only for helo - percent',
    convert: '{VAL} / 16384 * 100',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine2FuelPres: new Offset({
    value: 0x990,
    name: 'engine2FuelPres',
    category: OffsetCategory.ENGINE,
    description: 'engine2 fuel pressure - PSI',
    convert: '{VAL} / 144',
    type: Type.UInt32,
    permission: 'r',
  }),
  engine2FuelUsedSinceStart: new Offset({
    value: 0x9A4,
    name: 'engine2FuelUsedSinceStart',
    category: OffsetCategory.ENGINE,
    description: 'engine2 used fuel since start - lb',
    type: Type.Single,
    permission: 'r',
  }),
  engine2FuelElapsedTime: new Offset({
    value: 0x9A8,
    name: 'engine2FuelElapsedTime',
    category: OffsetCategory.ENGINE,
    description: 'engine2 fuel elapsed time - h',
    type: Type.Single,
    permission: 'r',
  }),
  engine2FuelFlowLbHour: new Offset({
    value: 0x9B0,
    name: 'engine2FuelFlowLbHour',
    category: OffsetCategory.ENGINE,
    description: 'engine2 fuel elapsed time - lb/h',
    type: Type.Double,
    permission: 'r',
  }),
  engine2TorqueFtLb: new Offset({
    value: 0x9B8,
    name: 'engine2TorqueFtLb',
    category: OffsetCategory.ENGINE,
    description: 'engine2 fuel elapsed time - ft lb',
    type: Type.Single,
    permission: 'r',
  }),
};
