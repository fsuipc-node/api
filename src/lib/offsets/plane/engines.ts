import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

import { engine1 } from './engines/engine1';
import { engine2 } from './engines/engine2';
import { engine3 } from './engines/engine3';
import { engine4 } from './engines/engine4';

export const engines: OffsetList = {
  engineType: new Offset({
    value: 0x609,
    name: 'engineType',
    category: OffsetCategory.ENGINE,
    description: 'engine type',
    convert: 'engineType',
    mapping: true,
    type: Type.Byte,
    permission: 'r',
  }),
  hasMixtureControl: new Offset({
    value: 0x780,
    name: 'hasMixtureControl',
    category: OffsetCategory.ENGINE,
    description: 'has mixture control',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  hasCarbHeat: new Offset({
    value: 0x784,
    name: 'hasCarbHeat',
    category: OffsetCategory.ENGINE,
    description: 'has carb heat',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  activeEngine: new Offset({
    value: 0x888,
    name: 'activeEngine',
    category: OffsetCategory.ENGINE,
    description: 'active engine pattern',
    type: Type.BitArray,
    length: 1,
    permission: 'rw',
  }),
  ...engine1,
  ...engine2,
  ...engine3,
  ...engine4,
};
