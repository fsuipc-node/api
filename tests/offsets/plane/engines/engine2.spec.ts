import { engine2 as offsets } from '@offsets/plane/engines/engine2';
import { replaceOffsetExpressionValue } from '@convert/apply-conversion';

describe('offset - plane/engines/engine2', () => {
  const offsetsTestCases = [
    { name: 'engine2ThrottleLever', value: -2048, expectedResult: -50 },
    { name: 'engine2ThrottleLever', value: 0, expectedResult: 0 },
    { name: 'engine2ThrottleLever', value: 8192, expectedResult: 50 },
    { name: 'engine2PropLever', value: -2048, expectedResult: -50 },
    { name: 'engine2PropLever', value: 0, expectedResult: 0 },
    { name: 'engine2PropLever', value: 8192, expectedResult: 50 },
    { name: 'engine2MixtureLever', value: 8192, expectedResult: 50 },
    { name: 'engine2Firing', value: 0, expectedResult: false },
    { name: 'engine2Firing', value: 1, expectedResult: true },
    { name: 'engine2N2', value: 8192, expectedResult: 50 },
    { name: 'engine2N1', value: 8192, expectedResult: 50 },
    { name: 'engine2PropRPM', value: 107374182400, expectedResult: 100 },
    { name: 'engine2FuelFlowLbHourSSL', value: 256, expectedResult: 2 },
    { name: 'engine2AntiIce', value: 0, expectedResult: false },
    { name: 'engine2AntiIce', value: 1, expectedResult: true },
    { name: 'engine2OilTemp', value: 10532, expectedResult: 90 },
    { name: 'engine2OilPres', value: 22342, expectedResult: 75 },
    { name: 'engine2PressureRatio', value: 20480, expectedResult: 2 },
    { name: 'engine2EGT', value: 14281, expectedResult: 750 },
    { name: 'engine2MP', value: 20480, expectedResult: 20 },
    { name: 'engine2OilQuantity', value: 8167, expectedResult: 50 },
    { name: 'engine2Vibration', value: 6553, expectedResult: 2 },
    { name: 'engine2HydPres', value: 64, expectedResult: 16 },
    { name: 'engine2HydQuantity', value: 8192, expectedResult: 50 },
    { name: 'engine2ITT', value: 3276800, expectedResult: 200 },
    { name: 'engine2Torque', value: 8192, expectedResult: 50 },
    { name: 'engine2FuelPres', value: 288, expectedResult: 2 },
  ];

  describe('offsets list', () => {
    it('should have required properties', () => {
      expect(offsets).toMatchSnapshot();
    });
  });

  offsetsTestCases.forEach(testedOffset => {
    describe(testedOffset.name, () => {
      it('should convert data properly', () => {
        const convertExpression = replaceOffsetExpressionValue(offsets[testedOffset.name], testedOffset.value);

        // tslint:disable-next-line:no-eval
        expect(eval(convertExpression)).toEqual(testedOffset.expectedResult);
      });
    });
  });
});
