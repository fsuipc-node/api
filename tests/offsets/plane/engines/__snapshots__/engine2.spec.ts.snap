// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`offset - plane/engines/engine2 offsets list should have required properties 1`] = `
Object {
  "engine2AntiIce": Offset {
    "category": "engine",
    "convert": "!!{VAL}",
    "description": "engine2 anti-ice/carb heat active",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2AntiIce",
    "permission": "r",
    "type": 5,
    "value": 2378,
  },
  "engine2CHT": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine2 cylinder head temperature - F",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2CHT",
    "permission": "r",
    "type": 8,
    "value": 2432,
  },
  "engine2EGT": Offset {
    "category": "engine",
    "convert": "Math.round({VAL} * 860 / 16384)",
    "description": "engine2 exhaust gas temperature - Trust only on jet engine - celsius",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2EGT",
    "permission": "r",
    "type": 2,
    "value": 2390,
  },
  "engine2Firing": Offset {
    "category": "engine",
    "convert": "!!{VAL}",
    "description": "engine 2 combustion",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2Firing",
    "permission": "r",
    "type": 5,
    "value": 2348,
  },
  "engine2FuelElapsedTime": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine2 fuel elapsed time - h",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2FuelElapsedTime",
    "permission": "r",
    "type": 9,
    "value": 2472,
  },
  "engine2FuelFlowLbHour": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine2 fuel elapsed time - lb/h",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2FuelFlowLbHour",
    "permission": "r",
    "type": 8,
    "value": 2480,
  },
  "engine2FuelFlowLbHourSSL": Offset {
    "category": "engine",
    "convert": "{VAL} / 128",
    "description": "engine2 fuel flow (lb per hour, sea level)",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2FuelFlowLbHourSSL",
    "permission": "r",
    "type": 5,
    "value": 2360,
  },
  "engine2FuelPres": Offset {
    "category": "engine",
    "convert": "{VAL} / 144",
    "description": "engine2 fuel pressure - PSI",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2FuelPres",
    "permission": "r",
    "type": 3,
    "value": 2448,
  },
  "engine2FuelUsedSinceStart": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine2 used fuel since start - lb",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2FuelUsedSinceStart",
    "permission": "r",
    "type": 9,
    "value": 2468,
  },
  "engine2HydPres": Offset {
    "category": "engine",
    "convert": "{VAL} / 4",
    "description": "engine2 hydraulic pressure - PSI",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2HydPres",
    "permission": "r",
    "type": 3,
    "value": 2416,
  },
  "engine2HydQuantity": Offset {
    "category": "engine",
    "convert": "{VAL} / 16384 * 100",
    "description": "engine2 hydraulic quantity - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2HydQuantity",
    "permission": "r",
    "type": 3,
    "value": 2420,
  },
  "engine2ITT": Offset {
    "category": "engine",
    "convert": "{VAL} / 16384",
    "description": "engine2 turbine temperature - C",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2ITT",
    "permission": "r",
    "type": 3,
    "value": 2440,
  },
  "engine2MP": Offset {
    "category": "engine",
    "convert": "{VAL} / 1024",
    "description": "engine2 manifold pressure - inHg",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2MP",
    "permission": "r",
    "type": 5,
    "value": 2392,
  },
  "engine2MixtureLever": Offset {
    "category": "engine",
    "convert": "{VAL} / 16384 * 100",
    "description": "engine 2 prop lever - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2MixtureLever",
    "permission": "rw",
    "type": 2,
    "value": 2344,
  },
  "engine2N1": Offset {
    "category": "engine",
    "convert": "{VAL} / 16384 * 100",
    "description": "engine 2 N1 - also helo RPM %",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2N1",
    "permission": "rw",
    "type": 2,
    "value": 2352,
  },
  "engine2N2": Offset {
    "category": "engine",
    "convert": "{VAL} / 16384 * 100",
    "description": "engine 2 N2 - also helo RPM %",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2N2",
    "permission": "rw",
    "type": 2,
    "value": 2350,
  },
  "engine2OilPres": Offset {
    "category": "engine",
    "convert": "Math.round({VAL} * 55 / 16384)",
    "description": "engine2 oil pressure - PSI",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2OilPres",
    "permission": "r",
    "type": 5,
    "value": 2386,
  },
  "engine2OilQuantity": Offset {
    "category": "engine",
    "convert": "Math.round({VAL} / 16384 * 100)",
    "description": "engine2 oil quantity - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2OilQuantity",
    "permission": "r",
    "type": 3,
    "value": 2408,
  },
  "engine2OilTemp": Offset {
    "category": "engine",
    "convert": "Math.round({VAL} * 140 / 16384)",
    "description": "engine2 oil temp - celsius",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2OilTemp",
    "permission": "r",
    "type": 2,
    "value": 2384,
  },
  "engine2PressureRatio": Offset {
    "category": "engine",
    "convert": "{VAL} * 1.6 / 16384",
    "description": "engine2 pressure ratio",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2PressureRatio",
    "permission": "r",
    "type": 2,
    "value": 2388,
  },
  "engine2PropLever": Offset {
    "category": "engine",
    "convert": "Math.round({VAL} < 0 ? {VAL} / 4096 * 100 : {VAL} / 16384 * 100)",
    "description": "engine 2 prop lever - percent - negative = reverse",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2PropLever",
    "permission": "rw",
    "type": 2,
    "value": 2342,
  },
  "engine2PropRPM": Offset {
    "category": "engine",
    "convert": "{VAL} / 16384 / 65536",
    "description": "derive RPM by multiplying by  - negative = counter-rotating propeller",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2PropRPM",
    "permission": "rw",
    "type": 2,
    "value": 2352,
  },
  "engine2RPMScaler": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine2 RPM Scaler - props: used to calculate RPM",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2RPMScaler",
    "permission": "r",
    "type": 2,
    "value": 2400,
  },
  "engine2StarterSwitchPosition": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine2 start switch position - JET: 0=off 1=start 2=gen/alt - PROP: 0=off 1=right 2=left 3=both 4=start",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2StarterSwitchPosition",
    "permission": "rw",
    "type": 5,
    "value": 2346,
  },
  "engine2ThrottleLever": Offset {
    "category": "engine",
    "convert": "Math.round({VAL} < 0 ? {VAL} / 4096 * 100 : {VAL} / 16384 * 100)",
    "description": "engine 2 throttle lever - percent - negative = reverse",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2ThrottleLever",
    "permission": "r",
    "type": 2,
    "value": 2340,
  },
  "engine2ThrottleLeverControl": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine2 throttle lever control - -4096 to +16384",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2ThrottleLeverControl",
    "permission": "rw",
    "type": 2,
    "value": 2354,
  },
  "engine2Torque": Offset {
    "category": "engine",
    "convert": "{VAL} / 16384 * 100",
    "description": "engine2 torque - trust only for helo - percent",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2Torque",
    "permission": "r",
    "type": 3,
    "value": 2444,
  },
  "engine2TorqueFtLb": Offset {
    "category": "engine",
    "convert": undefined,
    "description": "engine2 fuel elapsed time - ft lb",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2TorqueFtLb",
    "permission": "r",
    "type": 9,
    "value": 2488,
  },
  "engine2Vibration": Offset {
    "category": "engine",
    "convert": "Math.round({VAL} * 5 / 16384)",
    "description": "engine2 vibration",
    "length": undefined,
    "mapping": undefined,
    "name": "engine2Vibration",
    "permission": "r",
    "type": 3,
    "value": 2412,
  },
}
`;
